### ISSUE TITLE
[JENKINS-47897] Blue Ocean pipeline view shows stages and steps from previous build, until given step starts

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#4574]
<p>We are using Jenkinsfile.groovy to generate stages and steps for each build dynamically, based on parameters for given build.</p>

<p>(Actually, if it matters, in our case, from the UI point of view, stages are the same and the steps differ with names.)</p>

<p> </p>

<p>The problem is that then a new build starts Blue Ocean pipeline view shows it with stages and steps from the previous run!</p>

<p> </p>

<p>Only after given step start, its names is updated.</p>

<p> </p>

<p>I attach a screenshot showing that behaviour.</p>

<p> </p>

<p>Please always show the stages and steps from the current build, never use any cached values from the previous run as this is misleading.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/grzegorz_dubicki.png" title="grzegorz_dubicki's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=grzegorz_dubicki">grzegorz_dubicki</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897" target="_blank">Blue Ocean pipeline view shows stages and steps from previous build, until given step starts</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>label(s)</b>: pipeline, stage
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2018-02-19T00:39:44+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 10
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We are using Jenkinsfile.groovy to generate stages and steps for each build dynamically, based on parameters for given build.</p>

<p>(Actually, if it matters, in our case, from the UI point of view, stages are the same and the steps differ with names.)</p>

<p> </p>

<p>The problem is that then a new build starts Blue Ocean pipeline view shows it with stages and steps from the previous run!</p>

<p> </p>

<p>Only after given step start, its names is updated.</p>

<p> </p>

<p>I attach a screenshot showing that behaviour.</p>

<p> </p>

<p>Please always show the stages and steps from the current build, never use any cached values from the previous run as this is misleading.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins 2.89.1<br/>
Blue Ocean 1.3.5<br/>
(first observed in:<br/>
Jenkins 2.60.2<br/>
Blue Ocean 1.2.2)
```
</details>
<details><summary><i>1 attachment</i></summary>

- [Screen Shot 2017-11-08 at 14.52.36.png](https://issues.jenkins.io/secure/attachment/40264/Screen%20Shot%202017-11-08%20at%2014.52.36.png)
  > ![Screen Shot 2017-11-08 at 14.52.36.png](https://issues.jenkins.io/secure/attachment/40264/Screen%20Shot%202017-11-08%20at%2014.52.36.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/grzegorz_dubicki.png" title="grzegorz_dubicki's avatar" /> grzegorz_dubicki:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897?focusedId=322600&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-322600">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have upgraded Jenkins and Blue Ocean to latest versions (see Environment field) and still observe this issue.</p>
</pre>
</details>

</ul>
</details>

<p>I have upgraded Jenkins and Blue Ocean to latest versions (see Environment field) and still observe this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897?focusedId=329187&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329187">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=grzegorz_dubicki" class="user-hover" rel="grzegorz_dubicki">grzegorz_dubicki</a> yes, what you are seeing is desired behaviour for most users as many have reasonably static Pipeline elements (e.g. stages stay the same name, are not dynamically created). As we used a scripted language, we can't "read ahead" of the Pipeline without first executing the Pipeline, so we make assumptions that the Pipeline hasn't really changed since the previous run. In your case, we can't make that assumption due to the very dynamic nature of your Pipeline.</p>

<p>Ill be marking this ticket as "wont' do" because I am hesitant of changing this behaviour today as this has an unknown impact on other users. However, I would like to look at this again if we receive a few more reports of this issue. We can reopen this ticket when the time comes.</p>

<p>Thanks for reporting and sorry I was unable to help this time.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=grzegorz_dubicki" class="user-hover" rel="grzegorz_dubicki">grzegorz_dubicki</a> yes, what you are seeing is desired behaviour for most users as many have reasonably static Pipeline elements (e.g. stages stay the same name, are not dynamically created). As we used a scripted language, we can't "read ahead" of the Pipeline without first executing the Pipeline, so we make assumptions that the Pipeline hasn't really changed since the previous run. In your case, we can't make that assumption due to the very dynamic nature of your Pipeline.</p>

<p>Ill be marking this ticket as "wont' do" because I am hesitant of changing this behaviour today as this has an unknown impact on other users. However, I would like to look at this again if we receive a few more reports of this issue. We can reopen this ticket when the time comes.</p>

<p>Thanks for reporting and sorry I was unable to help this time.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>johnhus:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897?focusedId=329229&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329229">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Is there a way to clean the predefined steps? like a parameter / function in the script to clean the view of the previous run?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Is there a way to clean the predefined steps? like a parameter / function in the script to clean the view of the previous run?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897?focusedId=329332&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329332">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=johnhus" class="user-hover" rel="johnhus">johnhus</a> theres nothing like that right now. We'd be happy to accept a PR for that if you or someone else had the time. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=johnhus" class="user-hover" rel="johnhus">johnhus</a> theres nothing like that right now. We'd be happy to accept a PR for that if you or someone else had the time. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sabibo:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-47897?focusedId=403933&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403933">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We really need this also!</p>
</pre>
</details>

</ul>
</details>

<p>We really need this also!</p>
